### PR TITLE
MB-2142/MB-2661 Retrieve, parse, and return the most recent weekly national average diesel fuel price and publication date from the EIA Open Data API &  Set up save_ghc_diesel_fuel_price_data task

### PR DESCRIPTION
**Pull request by harshamurthy**
## Description

Fuel Surcharge is a service item for GHC moves. To calculate Fuel Surcharge, MilMove needs access to the most recent National Average Diesel Fuel Price published by the EIA.

According to Appendix G Section G.2 of the GHC proposal:

Fuel rate adjustments are based on the weekly U.S. No. 2 Diesel (On- Highway, All Types) Retail prices published on Monday by the Department of Energy (DOE) found at http://www.eia.gov

Due to these requirements, MilMove must store this data.

This PR is creates a new service `ghcdieselfuelprice` responsible for retrieving the weekly national average diesel fuel price from the EIA Open Data API, parsing the returned data to obtain the price and publication date, and returning that data. It also includes a task and CLI script, `save_ghc_diesel_fuel_price_data.go`, to run this `ghcdieselfuelprice` service.

## Setup
Pull this branch down locally and run the following:

- `make prune` to clear out Docker volumes

- `make db_dev_reset db_dev_migrate` to reset dev database

- `make db_test_reset db_test_migrate` to reset test database

- `rm bin/milmove-tasks` & `make bin/milmove-tasks` to rebuild all MilMove tasks including the one introduced by this PR

- `go test ./pkg/services/ghcdieselfuelprice` to run the tests for the `ghcdieselfuelprice` service

- `milmove-tasks save-ghc-diesel-fuel-price-data` to fetch the most recent fuel price data from the EIA API - you should see `{"last updated": "2020-06-22T18:16:52-0400", "publication date": "20200622", "price": 2.425}` in the logs

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/backend#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/backend#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/wiki/migrate-the-database#zero-downtime-migrations))
  * [ ] Have been communicated to #g-database
  * [ ] Secure migrations have been tested following the instructions in our [docs](https://github.com/transcom/mymove/wiki/migrate-the-database#secure-migrations)
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story MB-2142](https://dp3.atlassian.net/browse/MB-2142) for this change
* [Jira story MB-2661](https://dp3.atlassian.net/browse/MB-2661) for this change
* [Jira epic](https://dp3.atlassian.net/browse/MB-2648) for this change

### CODE REVIEW

File: pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher_test.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		dieselFuelPriceStorer := NewDieselFuelPriceStorer("https://api.eia.gov/series/", "pUW34B2q8tLooWEVQpU7s9Joq672q2rP", suite.helperStubEiaData)
+
+		eiaFinalURL, err := BuildEiaAPIURL(dieselFuelPriceStorer.eiaURL, dieselFuelPriceStorer.eiaKey)
+		suite.NoError(err)
+		dieselFuelPriceStorer.eiaFinalURL = eiaFinalURL
+
+		suite.Equal("https://api.eia.gov/series/?api_key=pUW34B2q8tLooWEVQpU7s9Joq672q2rP&series_id=PET.EMD_EPD2D_PTE_NUS_DPG.W", dieselFuelPriceStorer.eiaFinalURL)
+	})
+
+	// TODO: Figure out how to test FetchEiaData function without needing to make API call
```
[the remark this answers is not in the document]
Comment (harshamurthy) [reply pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher_test.go:18]: This test specifically deals with successfully retreiving a response from the EIA Open Data API. The rest of the tests use stubbed out response data.

I think it is important to test that the API call and the data returned from it are working properly, but also don't want to be making a live call in the tests.

Is there are standard way to approach this? Or is this something that is not really worth testing as long as any potential errors from the API being down or receiving an invalid response are handled correctly?

File: pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+	// TODO: Return an error if EiaFinalUrl is nil
+
+	response, err := client.Get(eiaFinalURL)
+	if err != nil {
+		return eiaData, fmt.Errorf("GET request to EIA Open Data API failed: %w", err)
+	}
+
+	eiaData.ResponseStatusCode = response.StatusCode
+	// TODO: Log Status Code
```
[the remark this answers is not in the document]
Comment (harshamurthy) [reply pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher.go:42]: I figured logging the status code would make it easier in a scenario where someone is trying to debug. May not be necessary though given errors are being handled and have descriptive messages.

File: pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (garrettqmartin8) [reply pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher_test.go:18]: I think you've approached it in the right way in handling as many potential errors/cases as you can without calling the actual API.

If you'd feel more comfortable having some more assurance that the real API behaves like you expect, you could use something like [`go-vcr`](https://github.com/dnaeon/go-vcr) which will make a real API call the first time you run the tests and record the response. Subsequent test runs will use the recorded response for speedy and realistic integration tests. I haven't used `go-vcr`, but I've used the Ruby gem that it's inspired by with success.

That being said, I think the coverage you have is good even without this test.

File: pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	query.Set("api_key", eiaKey)
+	query.Set("series_id", "PET.EMD_EPD2D_PTE_NUS_DPG.W")
+	parsedURL.RawQuery = query.Encode()
+	eiaFinalURL = parsedURL.String()
+
+	return eiaFinalURL, nil
+}
+
+// FetchEiaData makes a call to the EIA Open Data API and returns the API response
+func FetchEiaData(eiaFinalURL string) (EiaData, error) {
```
Comment (jacquelineIO) [inline pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher.go:30]: Is this not in use anywhere is it being tested?

Comment (harshamurthy) [reply pkg/services/ghcdieselfuelprice/ghc_diesel_fuel_price_fetcher.go:30]: This function makes the actual call to the API and needs the API key to run successfully. I had originally written a test for it, but decided it was not great practice to include a real API key in the test file even if it was just being used for the purpose of the test. I think it'll be best to test this function further down stream.

Also, as far as testing this code via the CLI, writing the task to run this code was not originally a part of this scope. But I have pulled in [MB-2661](https://dp3.atlassian.net/browse/MB-2661) into the sprint and can include that work in this PR if that makes sense.

**Pull request merged**